[Feature] Support dynamic modification of datacache.enable

stonechen111 opened the pull request:
Supports dynamic modification of table partition level datacache.enable attributes:
1. StarOSWorker uses heartbeat to report the datacache.enable of the shards it manages.
2. StarManager compares the shardInfoMd5 of ShardReportInfos. If they are not equal, it
   sends rpc to starlet to update the corresponding shard.

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [x] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

sduzh commented on be/src/storage/lake/schema_change.cpp:
```diff
@@ -348,8 +348,9 @@ Status SchemaChangeHandler::process_update_tablet_meta(const TUpdateTabletMetaIn
-    if (tablet_meta_info.meta_type != TTabletMetaType::ENABLE_PERSISTENT_INDEX) {
-        // Only support ENABLE_PERSISTENT_INDEX for now
+    if (tablet_meta_info.meta_type != TTabletMetaType::ENABLE_PERSISTENT_INDEX
+            || tablet_meta_info.meta_type != TTabletMetaType::DATACACHE_ENABLE) {
```
```suggestion
            && tablet_meta_info.meta_type != TTabletMetaType::DATACACHE_ENABLE) {
```

sduzh commented on fe/fe-core/src/main/java/com/starrocks/alter/LakeTableAlterMetaJob.java:
```diff
@@ -205,6 +205,11 @@ protected void runFinishedRewritingJob() throws AlterCancelException {
                     tempProperties.put(PropertyAnalyzer.PROPERTIES_ENABLE_PERSISTENT_INDEX, String.valueOf(metaValue));
                     GlobalStateMgr.getCurrentState().getLocalMetastore()
                             .modifyTableMeta(db, table, tempProperties, metaType);
+                } else if (metaType == TTabletMetaType.DATACACHE_ENABLE) {
```
Only need to update the shard info in StarMgr, no need to create this `LakeTableAlterMetaJob`

sduzh commented on be/src/service/staros_worker.h:
```diff
@@ -53,6 +53,8 @@ class StarOSWorker : public staros::starlet::Worker {
 
     absl::Status add_shard(const ShardInfo& shard) override;
 
+    absl::Status modify_shard(const ShardId shard, const bool enable_cache) override;
```
Question: when will this method been invoked?

stonechen111 replied:
fe -> starManager: alterShard
starManager -> StarOS_worker: modifyShard

sduzh replied:
The function name of modify_shard() was too generic, make it more precisely capture what the function does.

kevincai commented on fe/fe-core/src/main/java/com/starrocks/alter/AlterJobMgr.java:
```diff
@@ -946,8 +949,19 @@ public void modifyPartitionsProperty(Database db,
+            if (olapTable.isCloudNativeTable()) {
+                DataCacheInfo dataCacheInfo = partitionInfo.getDataCacheInfo(partition.getId());
+                if (dataCacheInfo == null || newEnableDataCache != dataCacheInfo.isEnabled()) {
+                    GlobalStateMgr.getCurrentState().getStarOSAgent().updateShardGroup(
```
may need to keep this change as close as the ModifyPartitionInfo persistency, otherwise the metadata can get easily out of sync, for example, the updateShardGroup is success, but this modification is failed at last. The system is in an inconsistent state.

stonechen111 replied:
OK, I'll keep them as close as possible.

kevincai commented on fe/fe-core/src/main/java/com/starrocks/catalog/TableProperty.java:
```diff
@@ -216,6 +216,8 @@ public static String valueList() {
 
     private PeriodDuration dataCachePartitionDuration;
 
+    private Boolean dataCacheEnable;
```
looks like this variable is not used at all.

stonechen111 replied:
The variables is used in the old version, I forgot to remove it, I'll remove it.

kevincai commented on fe/fe-core/src/main/java/com/starrocks/common/util/PropertyAnalyzer.java:
```diff
@@ -1111,6 +1111,12 @@ public static PeriodDuration analyzeDataCachePartitionDuration(Map<String, Strin
     }
 
+    public static boolean analyzeDataCacheEnable(Map<String, String> properties) throws AnalysisException {
+        boolean enableDataCache = analyzeBooleanProp(properties, PropertyAnalyzer.PROPERTIES_DATACACHE_ENABLE, true);
+
+        return enableDataCache;
```
```suggestion
        return analyzeBooleanProp(properties, PropertyAnalyzer.PROPERTIES_DATACACHE_ENABLE, true);
```

stonechen111 replied:
OK

kevincai commented on fe/fe-core/src/main/java/com/starrocks/lake/StarOSAgent.java:
```diff
@@ -475,6 +479,39 @@ public List<Long> createShards(int numShards, FilePathInfo pathInfo, FileCacheIn
+                                .build());
+            }
+
+            client.updateShardGroup(serviceId, updateShardGroupInfoList);
```
consider what to do with the partial success case.

stonechen111 replied:
If any exception occurs during the updateShardGroup process, all shards will be rolled back to their original state, so there is no partial sucess.

kevincai commented on fe/fe-core/src/main/java/com/starrocks/lake/StorageInfo.java:
```diff
@@ -59,6 +59,15 @@ public DataCacheInfo getDataCacheInfo() {
     }
 
+    public void setDataCacheEnable(boolean isEnable) {
+        FileCacheInfo newCacheInfo = FileCacheInfo.newBuilder()
+                .setEnableCache(isEnable)
+                .setTtlSeconds(cacheInfo.getTtlSeconds())
+                .setAsyncWriteBack(cacheInfo.getAsyncWriteBack()).build();
+
+        this.cacheInfo = newCacheInfo;
```
```suggestion
        this.cacheInfo = this.cacheInfo.toBuilder().setEnableCache(isEnabled).build();
```

stonechen111 replied:
OK

kevincai commented on fe/fe-core/src/main/java/com/starrocks/persist/ModifyPartitionInfo.java:
```diff
@@ -137,6 +144,7 @@ public void write(DataOutput out) throws IOException {
 
         out.writeShort(replicationNum);
         out.writeBoolean(isInMemory);
+        out.writeBoolean(dataCacheEnable);
```
can't do it like this. double check its compatibility, suppose we should not do any change to write() and readFields(). should only leverage gson serialize to persistent attributes.

stonechen111 replied:
Discuss it with you face to face tomorrow

kevincai commented on fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:
```diff
@@ -4160,7 +4164,42 @@ public void alterTableProperties(Database db, OlapTable table, Map<String, Strin
+                    LakeTable lakeTable = (LakeTable) table;
+                    GlobalStateMgr.getCurrentState().getStarOSAgent().updateShardGroup(lakeTable, isEnable);
+                } else {
+                    throw new DdlException("Only support alter datacache.enable for lake tables");
```
there is no lake tables to end user, check `ErrorReportException`, if we can add new error code or reuse similar error code.

stonechen111 replied:
Discuss it with you face to face tomorrow

kevincai commented on fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:
```diff
@@ -4160,7 +4164,42 @@ public void alterTableProperties(Database db, OlapTable table, Map<String, Strin
 
+                if (table instanceof LakeTable) {
+                    LakeTable lakeTable = (LakeTable) table;
+                    GlobalStateMgr.getCurrentState().getStarOSAgent().updateShardGroup(lakeTable, isEnable);
```
this will throw StarClientException, what will end user see if excepted here?

stonechen111 replied:
updateShardGroup will throw DdlException if error occurs.
The following picture shows what happen in shared-nothing mode if we want to alter datacache.enable
![image](https://github.com/StarRocks/starrocks/assets/145319350/f1781b2c-f437-47b4-ae2c-3a8f75570652)

kevincai commented on fe/fe-core/src/test/java/com/starrocks/alter/LakeTableAlterDataCachePartitionDurationTest.java:
why removed?

stonechen111 replied:
The Test file is only used for datacache.partition_duration. There is no need to create a separate test file for a single property, these test cases have been merged into LakeTableAlterMetaJobTest along with datacache.enable.

kevincai commented on fe/fe-core/src/test/java/com/starrocks/lake/StarOSAgentTest.java:
```diff
@@ -673,4 +675,26 @@ public List<WorkerGroupDetailInfo> listWorkerGroup(String serviceId, List<Long>
+        table.addPartition(partition1);
+        table.addPartition(partition2);
+
+        try {
```
use `ExceptionChecker.expectThrowsNoException`

stonechen111 replied:
OK

stonechen111 commented on be/src/service/staros_worker.cpp:
```diff
@@ -81,8 +81,7 @@ StarOSWorker::~StarOSWorker() = default;
     auto it = _shards.find(shard.id);
-    if (it != _shards.end() && it->second.shard_info.path_info.has_fs_info() && shard.path_info.has_fs_info() &&
-        it->second.shard_info.path_info.fs_info().version() < shard.path_info.fs_info().version()) {
+    if (it != _shards.end() && it->second.shard_info.hash_code != shard.hash_code) {
```
New starmgr can handle hearbeat from old starlet so shown below:
![image](https://github.com/StarRocks/starrocks/assets/145319350/fb76aa41-108d-479d-a3c1-675d22e1d2b4)
When old starlet receive heartbeat from new starmgr, the hash code will be 0 as shown below:
![image](https://github.com/StarRocks/starrocks/assets/145319350/77cf8762-732a-46f1-94f9-81583f1c3aab)
so, we can use hash_code to tell new starmgr and old starmgr apart as shown below:
![image](https://github.com/StarRocks/starrocks/assets/145319350/32c4f0bb-22ba-4c8b-a542-337bd3bbbc4f)

stonechen111 commented on fe/fe-core/src/main/java/com/starrocks/alter/AlterJobMgr.java:
```diff
@@ -946,8 +949,19 @@ public void modifyPartitionsProperty(Database db,
                 partitionInfo.setTabletType(partition.getId(), tTabletType);
             }
+            // 5. enable data cache
+            if (olapTable.isCloudNativeTable()) {
```
This property is also allowed in MV, I will change isCloudNativeTable to  isCloudNativeTableOrMaterializedView.

stonechen111 committed:
[Feature] Support dynamic modification of datacache.enable

Supports dynamic modification of table and partition level datacache.enable attributes:
1. Synchronously modify the datacache.enable of the shardGroup in StarManager through
   StarManager's updateShardGroup rpc.
2. Through the heartbeat detection of the starlet worker, asynchronously
   modify the cache enable status of the starlet worker.

Signed-off-by: stonechen111 <chenshi@starrocks.com>

kevincai commented on fe/fe-core/src/main/java/com/starrocks/lake/StarOSAgent.java:
```diff
@@ -475,6 +478,27 @@ public List<Long> createShards(int numShards, FilePathInfo pathInfo, FileCacheIn
+            for (Partition partition : partitionsList) {
+                partitionNamesBuilder.append(partition.getName()).append(" ");
+            }
+            throw new DdlException("Failed to partition alter partition shardGroups, PartitionNames: " +
```
fix the error message.

stonechen111 replied:
OK

kevincai commented on fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:
```diff
@@ -4160,7 +4164,50 @@ public void alterTableProperties(Database db, OlapTable table, Map<String, Strin
+                        List<Partition> partitionsList = new ArrayList<>(partitions);
+                        GlobalStateMgr.getCurrentState().getStarOSAgent().updateShardGroup(partitionsList, isEnable);
+                    } else {
+                        throw new DdlException("Only support alter datacache.enable for lake tables");
```
fix the error message. e.g. "property 'datacache.enable' is only supported in shared-data mode"

stonechen111 replied:
OK

kevincai commented on fe/fe-core/src/test/java/com/starrocks/alter/LakeTableAlterMetaJobTest.java:
start a new test file, use utframeUtils to start a real shared-data cluster, and run the alter property sql to test the end-to-end workflow.

stonechen111 replied:
OK

kevincai commented on fe/fe-core/src/test/java/com/starrocks/lake/StarOSAgentTest.java:
```diff
@@ -673,4 +675,32 @@ public List<WorkerGroupDetailInfo> listWorkerGroup(String serviceId, List<Long>
     }
+
+    @Test
+    public void testUpdateShardGroup() throws StarClientException, DdlException, UserException {
```
this case is just for code coverage, no valuable test point at all.

stonechen111 replied:
OK I'll remove it later.

stonechen111 commented on fe/fe-core/src/main/java/com/starrocks/alter/AlterJobMgr.java:
```diff
@@ -857,6 +859,10 @@ public void modifyPartitionsProperty(Database db,
         if (olapTable.getState() != OlapTableState.NORMAL) {
             throw InvalidOlapTableStateException.of(olapTable.getState(), olapTable.getName());
         }
+        if (!olapTable.isCloudNativeTableOrMaterializedView() &&
```
OK

stonechen111 commented on be/src/service/staros_worker.cpp:
```diff
@@ -81,8 +81,11 @@ StarOSWorker::~StarOSWorker() = default;
-        it->second.shard_info.path_info.fs_info().version() < shard.path_info.fs_info().version()) {
+    // shard.hash_code is 0 means this heartbeat is sent from old starmgr.
+    if (it != _shards.end() &&
+        ((shard.hash_code != 0 && it->second.shard_info.hash_code != shard.hash_code) ||
```
OK

kevincai commented on fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:
```diff
@@ -4165,7 +4169,49 @@ public void alterTableProperties(Database db, OlapTable table, Map<String, Strin
+                boolean isEnable = Boolean.parseBoolean(dataCacheEnable);
 
+                Locker locker = new Locker();
+                locker.lockDatabase(db, LockType.WRITE);
```
@wyb please help to evaluate if this db lock is appropriate.

kevincai reviewed (approved):
mostly in a

starrocks-xupeng commented on fe/fe-core/src/main/java/com/starrocks/alter/AlterJobMgr.java:
```diff
@@ -859,6 +861,10 @@ public void modifyPartitionsProperty(Database db,
         if (olapTable.getState() != OlapTableState.NORMAL) {
             throw InvalidOlapTableStateException.of(olapTable.getState(), olapTable.getName());
         }
+        if (properties.containsKey(PropertyAnalyzer.PROPERTIES_DATACACHE_ENABLE) &&
```
can we move this logic to LakeTableAlterMetaJobXXX?

starrocks-xupeng commented on fe/fe-core/src/main/java/com/starrocks/lake/StarOSAgent.java:
```diff
@@ -476,6 +479,27 @@ public List<Long> createShards(int numShards, FilePathInfo pathInfo, FileCacheIn
+            for (Partition partition : partitionsList) {
+                partitionNamesBuilder.append(partition.getName()).append(" ");
+            }
+            throw new DdlException("Failed to alter partition shardGroups, PartitionNames: " +
```
the message might be too big, maybe just print the first one?

starrocks-xupeng commented on fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:
```diff
@@ -4165,7 +4169,49 @@ public void alterTableProperties(Database db, OlapTable table, Map<String, Strin
+                    List<ModifyPartitionInfo> modifyPartitionInfos = Lists.newArrayList();
+                    for (Partition partition : partitions) {
+                        DataCacheInfo dataCacheInfo = partitionInfo.getDataCacheInfo(partition.getId());
+                        boolean asyncWriteBack = dataCacheInfo == null ? false : dataCacheInfo.isAsyncWriteBack();
```
asyncWriteBack this check should be removed
